tests/late_joiners.py

```diff
@@ -104,107 +107,106 @@ def run_requests(nodes, total_requests, start_id, final_msg, final_msg_id):
                 node_id += 1
                 c = clients[node_id % len(clients)]
                 try:
-                    c.rpc("LOG_record", {"id": id, "msg": long_msg})
+                    check_commit(
+                        c.rpc("LOG_record", {"id": id, "msg": long_msg}), result=True
+                    )
                 except (TimeoutError, requests.exceptions.ReadTimeout,) as e:
-                    LOG.info("Trying to access a suspended node")
+                    LOG.info("Trying to access a suspended network")
+                    if cant_fail:
+                        raise RuntimeError(e)
                 id += 1
+
         wait_for_nodes(nodes, final_msg, final_msg_id)
 
 
 def run(args):
-    hosts = ["localhost", "localhost", "localhost"]
+    hosts = ["localhost", "localhost", "localhost", "localhost"]
 
     with infra.ccf.network(
         hosts, args.binary_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
     ) as network:
         network.start_and_join(args)
         first_node, backups = network.find_nodes()
         all_nodes = network.get_joined_nodes()
-
         term_info = find_primary(network)
         first_msg = "Hello, world!"
         second_msg = "Hello, world hello!"
+        catchup_msg = "Hey world!"
         final_msg = "Goodbye, world!"
 
-        nodes_to_kill = [backups[0]]
-        nodes_to_keep = [first_node, backups[1]]
-
-        # first timer determines after how many seconds each node will be suspended
-        if not args.skip_suspension:
-            timeouts = []
-            for node in all_nodes:
-                t = random.uniform(1, 10)
-                LOG.info(f"Initial timer for node {node.node_id} is {t} seconds...")
-                timeouts.append((t, node))
-
-            for t, node in timeouts:
-                tm = Timer(t, timeout, args=[node, True, args.election_timeout / 1000],)
-                tm.start()
-
-        LOG.info(
-            "Adding another node after f = 0 but before we need to send append entries"
-        )
-        # check that a new node can catch up naturally
-        new_node = network.create_and_trust_node(
-            lib_name=args.package, host="localhost", args=args,
-        )
-        assert new_node
-        nodes_to_keep.append(new_node)
-
         with first_node.node_client() as mc:
             check_commit = infra.checker.Checker(mc)
             check = infra.checker.Checker()
 
-            run_requests(all_nodes, TOTAL_REQUESTS, 0, first_msg, 1000)
+            run_requests(all_nodes, int(TOTAL_REQUESTS / 2), 0, first_msg, 1000)
```

Comment:
I am strongly for re-using the existing `logging_app.py` to issue requests and make sure they are committed properly. It would significantly reduce the size of the file, avoid keeping track of how many requests were sent before (3rd argument to `run_requests`) and contribute to a richer `LoggingTxs` for the rest of our e2e infra (as I believe `verify()` would have to be updated slightly to check entries on one node only). Happy to delay it to a later PR but I believe that it should be a very quick job.

Reply:
There is already a ticket for this, you can add further comments there if you would like (https://github.com/microsoft/CCF/issues/769)

Reply:
I forgot we had this ticket! It's scheduled for the second PBFT milestone though. I am not keen on accumulating technical debt in our tests (and anywhere else!), especially that we will re-use this test for Raft once the safety issue is fixed. Can we schedule it earlier? I'm happy to do the change myself too 😃

Reply:
Next after this PR are #762 and #824 as far as I am concerned. I think prioritization of tickets can be discussed offline? I don't think anything else is going to be scheduled to go into the experimental pbft milestone however. You are welcome to pick up any tickets of course :)